File: android/app/src/main/java/com/kordar/ghostchat/models/ControlMessage.kt
package com.kordar.ghostchat.models

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.int
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long

/**
 * Application-level control messages over the encrypted DataChannel.
 *
 * Wire format: `{ "_ctrl": true, "type": "<wire>", <payload…> }`.
 * MUST stay byte-for-byte compatible with iOS `ControlMessage.swift`.
 */
sealed class ControlMessage {

    abstract val wireType: String

    data class Renegotiate(val sdp: String) : ControlMessage() {
        override val wireType: String = "renegotiate"
    }

    data object CallRequest : ControlMessage() {
        override val wireType: String = "call-request"
    }

    data class CallResponse(val accepted: Boolean) : ControlMessage() {
        override val wireType: String = "call-response"
    }

    data object CallEnd : ControlMessage() {
        override val wireType: String = "call-end"
    }

    data class SecurityAlert(val alert: String) : ControlMessage() {
        override val wireType: String = "security-alert"
    }

    data class MessageAck(val counter: Long) : ControlMessage() {
        override val wireType: String = "message-ack"
    }

    data class MessageRead(val counter: Long) : ControlMessage() {
        override val wireType: String = "message-read"
    }

    data object Ready : ControlMessage() {
        override val wireType: String = "ready"
    }

    data class PushToken(val token: String) : ControlMessage() {
        override val wireType: String = "push-token"
    }

    data class NotifyToken(val token: String) : ControlMessage() {
        override val wireType: String = "notify-token"
    }

    data class Typing(val isTyping: Boolean) : ControlMessage() {
        override val wireType: String = "typing"
    }

    data class Capabilities(val features: List<String>) : ControlMessage() {
        override val wireType: String = "capabilities"
    }

    data class FileStart(
        val fileId: String,
        val name: String,
        val size: Int,
        val mimeType: String,
        val totalChunks: Int
    ) : ControlMessage() {
        override val wireType: String = "file-start"
    }

    data class FileChunk(
        val fileId: String,
        val index: Int,
        val data: String
    ) : ControlMessage() {
        override val wireType: String = "file-chunk"
    }

    data class FileComplete(val fileId: String) : ControlMessage() {
        override val wireType: String = "file-complete"
    }

    data class FileRetransmit(val fileId: String, val indices: List<Int>) : ControlMessage() {
        override val wireType: String = "file-retransmit"
    }

    data class MessageDelete(val messageId: String) : ControlMessage() {
        override val wireType: String = "message-delete"
    }

    data class MessageEdit(val messageId: String, val newText: String) : ControlMessage() {
        override val wireType: String = "message-edit"
    }

    data class MessagePin(val messageId: String, val pinned: Boolean) : ControlMessage() {
        override val wireType: String = "message-pin"
    }

    sealed class DecodingError : RuntimeException() {
        data object MissingCtrlMarker : DecodingError()
        data class UnknownType(val type: String) : DecodingError()
        data class MissingField(val field: String) : DecodingError()
    }

    companion object {
        private val json = Json { ignoreUnknownKeys = true; encodeDefaults = false }

        /** Canonical JSON encoder — field order matches iOS where possible. */
        fun encode(message: ControlMessage): String =
            json.encodeToString(JsonElement.serializer(), toJsonObject(message))

        fun decode(raw: String): ControlMessage {
            val obj = json.parseToJsonElement(raw).jsonObject
            return fromJsonObject(obj)
        }

        fun toJsonObject(message: ControlMessage): JsonObject = buildJsonObject {
            put("_ctrl", JsonPrimitive(true))
            put("type", JsonPrimitive(message.wireType))
            when (message) {
                is Renegotiate   -> put("sdp", JsonPrimitive(message.sdp))
                is CallRequest, is CallEnd, is Ready -> Unit
                is CallResponse  -> put("accepted", JsonPrimitive(message.accepted))
                is SecurityAlert -> put("alert", JsonPrimitive(message.alert))
                is MessageAck    -> put("c", JsonPrimitive(message.counter))
                is MessageRead   -> put("c", JsonPrimitive(message.counter))
                is PushToken     -> put("token", JsonPrimitive(message.token))
                is NotifyToken   -> put("token", JsonPrimitive(message.token))
                is Typing        -> put("isTyping", JsonPrimitive(message.isTyping))
                is Capabilities  -> put("features", JsonArray(message.features.map { JsonPrimitive(it) }))
                is FileStart     -> {
                    put("fileId",      JsonPrimitive(message.fileId))
                    put("name",        JsonPrimitive(message.name))
                    put("size",        JsonPrimitive(message.size))
                    put("mimeType",    JsonPrimitive(message.mimeType))
                    put("totalChunks", JsonPrimitive(message.totalChunks))
                }
                is FileChunk -> {
                    put("fileId", JsonPrimitive(message.fileId))
                    put("index",  JsonPrimitive(message.index))
                    put("data",   JsonPrimitive(message.data))
                }
                is FileComplete    -> put("fileId", JsonPrimitive(message.fileId))
                is FileRetransmit  -> {
                    put("fileId",  JsonPrimitive(message.fileId))
                    put("indices", JsonArray(message.indices.map { JsonPrimitive(it) }))
                }
                is MessageDelete   -> put("messageId", JsonPrimitive(message.messageId))
                is MessageEdit     -> {
                    put("messageId", JsonPrimitive(message.messageId))
                    put("newText",   JsonPrimitive(message.newText))
                }
                is MessagePin      -> {
                    put("messageId", JsonPrimitive(message.messageId))
                    put("pinned",    JsonPrimitive(message.pinned))
                }
            }
        }

        fun fromJsonObject(obj: JsonObject): ControlMessage {
            val isCtrl = obj["_ctrl"]?.jsonPrimitive?.booleanOrNull ?: false
            if (!isCtrl) throw DecodingError.MissingCtrlMarker
            val type = obj["type"]?.jsonPrimitive?.content
                ?: throw DecodingError.MissingField("type")
            return when (type) {
                "renegotiate"      -> Renegotiate(obj.string("sdp"))
                "call-request"     -> CallRequest
                "call-response"    -> CallResponse(obj.bool("accepted"))
                "call-end"         -> CallEnd
                "security-alert"   -> SecurityAlert(obj.string("alert"))
                "message-ack"      -> MessageAck(obj.long("c"))
                "message-read"     -> MessageRead(obj.long("c"))
                "ready"            -> Ready
                "push-token"       -> PushToken(obj.string("token"))
                "notify-token"     -> NotifyToken(obj.string("token"))
                "typing"           -> Typing(obj.bool("isTyping"))
                "capabilities"     -> Capabilities(obj.stringList("features"))
                "file-start"       -> FileStart(
                    fileId      = obj.string("fileId"),
                    name        = obj.string("name"),
                    size        = obj.int("size"),
                    mimeType    = obj.string("mimeType"),
                    totalChunks = obj.int("totalChunks")
                )
                "file-chunk"       -> FileChunk(
                    fileId = obj.string("fileId"),
                    index  = obj.int("index"),
                    data   = obj.string("data")
                )
                "file-complete"    -> FileComplete(obj.string("fileId"))
                "file-retransmit"  -> FileRetransmit(
                    fileId  = obj.string("fileId"),
                    indices = obj.intList("indices")
                )
                "message-delete"   -> MessageDelete(obj.string("messageId"))
                "message-edit"     -> MessageEdit(obj.string("messageId"), obj.string("newText"))
                "message-pin"      -> MessagePin(obj.string("messageId"), obj.bool("pinned"))
                else               -> throw DecodingError.UnknownType(type)
            }
        }

        private fun JsonObject.string(key: String): String =
            this[key]?.jsonPrimitive?.content ?: throw DecodingError.MissingField(key)

        private fun JsonObject.bool(key: String): Boolean =
            this[key]?.jsonPrimitive?.boolean ?: throw DecodingError.MissingField(key)

        private fun JsonObject.int(key: String): Int =
            this[key]?.jsonPrimitive?.int ?: throw DecodingError.MissingField(key)

        private fun JsonObject.long(key: String): Long =
            this[key]?.jsonPrimitive?.long ?: throw DecodingError.MissingField(key)

        private fun JsonObject.stringList(key: String): List<String> =
            (this[key]?.jsonArray ?: throw DecodingError.MissingField(key))
                .map { it.jsonPrimitive.content }

        private fun JsonObject.intList(key: String): List<Int> =
            (this[key]?.jsonArray ?: throw DecodingError.MissingField(key))
                .map { it.jsonPrimitive.int }
    }
}

File: ios/GhostChat/Models/ControlMessage.swift
import Foundation

/// Application-level control messages sent over the encrypted DataChannel.
/// Wire format: `{ "_ctrl": true, "type": "<type>", <payload-fields...> }`.
/// Must remain byte-for-byte compatible with Android counterpart.
enum ControlMessage: Equatable {
    case renegotiate(sdp: String)
    case callRequest
    case callResponse(accepted: Bool)
    case callEnd
    case securityAlert(alert: String)
    case messageAck(counter: UInt64)
    case messageRead(counter: UInt64)
    case ready
    case pushToken(token: String)
    case notifyToken(token: String)
    case typing(isTyping: Bool)
    case capabilities(features: [String])
    case fileStart(fileId: String, name: String, size: Int, mimeType: String, totalChunks: Int)
    case fileChunk(fileId: String, index: Int, data: String)
    case fileComplete(fileId: String, sha256: String)
    case fileRetransmit(fileId: String, indices: [Int])
    case messageDelete(messageId: String)
    case messageEdit(messageId: String, newText: String)
    case messagePin(messageId: String, pinned: Bool)
}

extension ControlMessage {
    enum DecodingError: Error, Equatable {
        case missingCtrlMarker
        case unknownType(String)
    }

    /// Wire tag for each case. Keep in sync with Android `ControlMessageType`.
    var wireType: String {
        switch self {
        case .renegotiate:      return "renegotiate"
        case .callRequest:      return "call-request"
        case .callResponse:     return "call-response"
        case .callEnd:          return "call-end"
        case .securityAlert:    return "security-alert"
        case .messageAck:       return "message-ack"
        case .messageRead:      return "message-read"
        case .ready:            return "ready"
        case .pushToken:        return "push-token"
        case .notifyToken:      return "notify-token"
        case .typing:           return "typing"
        case .capabilities:     return "capabilities"
        case .fileStart:        return "file-start"
        case .fileChunk:        return "file-chunk"
        case .fileComplete:     return "file-complete"
        case .fileRetransmit:   return "file-retransmit"
        case .messageDelete:    return "message-delete"
        case .messageEdit:      return "message-edit"
        case .messagePin:       return "message-pin"
        }
    }
}

extension ControlMessage: Codable {
    private enum Keys: String, CodingKey {
        case ctrl = "_ctrl"
        case type
        case sdp, accepted, alert, c, token, isTyping, features
        case fileId, name, size, mimeType, totalChunks
        case index, data, indices, sha256
        case messageId, newText, pinned
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: Keys.self)
        let isCtrl = (try? c.decode(Bool.self, forKey: .ctrl)) ?? false
        guard isCtrl else { throw DecodingError.missingCtrlMarker }
        let type = try c.decode(String.self, forKey: .type)
        switch type {
        case "renegotiate":
            self = .renegotiate(sdp: try c.decode(String.self, forKey: .sdp))
        case "call-request":
            self = .callRequest
        case "call-response":
            self = .callResponse(accepted: try c.decode(Bool.self, forKey: .accepted))
        case "call-end":
            self = .callEnd
        case "security-alert":
            self = .securityAlert(alert: try c.decode(String.self, forKey: .alert))
        case "message-ack":
            self = .messageAck(counter: try c.decode(UInt64.self, forKey: .c))
        case "message-read":
            self = .messageRead(counter: try c.decode(UInt64.self, forKey: .c))
        case "ready":
            self = .ready
        case "push-token":
            self = .pushToken(token: try c.decode(String.self, forKey: .token))
        case "notify-token":
            self = .notifyToken(token: try c.decode(String.self, forKey: .token))
        case "typing":
            self = .typing(isTyping: try c.decode(Bool.self, forKey: .isTyping))
        case "capabilities":
            self = .capabilities(features: try c.decode([String].self, forKey: .features))
        case "file-start":
            self = .fileStart(
                fileId:      try c.decode(String.self, forKey: .fileId),
                name:        try c.decode(String.self, forKey: .name),
                size:        try c.decode(Int.self,    forKey: .size),
                mimeType:    try c.decode(String.self, forKey: .mimeType),
                totalChunks: try c.decode(Int.self,    forKey: .totalChunks)
            )
        case "file-chunk":
            self = .fileChunk(
                fileId: try c.decode(String.self, forKey: .fileId),
                index:  try c.decode(Int.self,    forKey: .index),
                data:   try c.decode(String.self, forKey: .data)
            )
        case "file-complete":
            self = .fileComplete(
                fileId: try c.decode(String.self, forKey: .fileId),
                sha256: try c.decode(String.self, forKey: .sha256)
            )
        case "file-retransmit":
            self = .fileRetransmit(
                fileId:  try c.decode(String.self, forKey: .fileId),
                indices: try c.decode([Int].self,  forKey: .indices)
            )
        case "message-delete":
            self = .messageDelete(messageId: try c.decode(String.self, forKey: .messageId))
        case "message-edit":
            self = .messageEdit(
                messageId: try c.decode(String.self, forKey: .messageId),
                newText:   try c.decode(String.self, forKey: .newText)
            )
        case "message-pin":
            self = .messagePin(
                messageId: try c.decode(String.self, forKey: .messageId),
                pinned:    try c.decode(Bool.self,   forKey: .pinned)
            )
        default:
            throw DecodingError.unknownType(type)
        }
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: Keys.self)
        try c.encode(true, forKey: .ctrl)
        try c.encode(wireType, forKey: .type)
        switch self {
        case .renegotiate(let sdp):
            try c.encode(sdp, forKey: .sdp)
        case .callRequest, .callEnd, .ready:
            break
        case .callResponse(let accepted):
            try c.encode(accepted, forKey: .accepted)
        case .securityAlert(let alert):
            try c.encode(alert, forKey: .alert)
        case .messageAck(let counter), .messageRead(let counter):
            try c.encode(counter, forKey: .c)
        case .pushToken(let token), .notifyToken(let token):
            try c.encode(token, forKey: .token)
        case .typing(let isTyping):
            try c.encode(isTyping, forKey: .isTyping)
        case .capabilities(let features):
            try c.encode(features, forKey: .features)
        case .fileStart(let fileId, let name, let size, let mimeType, let totalChunks):
            try c.encode(fileId,      forKey: .fileId)
            try c.encode(name,        forKey: .name)
            try c.encode(size,        forKey: .size)
            try c.encode(mimeType,    forKey: .mimeType)
            try c.encode(totalChunks, forKey: .totalChunks)
        case .fileChunk(let fileId, let index, let data):
            try c.encode(fileId, forKey: .fileId)
            try c.encode(index,  forKey: .index)
            try c.encode(data,   forKey: .data)
        case .fileComplete(let fileId, let sha256):
            try c.encode(fileId, forKey: .fileId)
            try c.encode(sha256, forKey: .sha256)
        case .fileRetransmit(let fileId, let indices):
            try c.encode(fileId,  forKey: .fileId)
            try c.encode(indices, forKey: .indices)
        case .messageDelete(let messageId):
            try c.encode(messageId, forKey: .messageId)
        case .messageEdit(let messageId, let newText):
            try c.encode(messageId, forKey: .messageId)
            try c.encode(newText,   forKey: .newText)
        case .messagePin(let messageId, let pinned):
            try c.encode(messageId, forKey: .messageId)
            try c.encode(pinned,    forKey: .pinned)
        }
    }
}
